queries/just/highlights.scm

```diff
@@ -0,0 +1,102 @@
+
+(comment) @comment
+(string) @string
+(boolean ["true" "false"]) @boolean
+["if" "else"] @conditional
+["export" "alias" "set"] @keyword
+["@" "==" "!=" "+" ":="] @operator
+[ "(" ")" "[" "]" "{{" "}}" "{" "}"] @punctuation.bracket
+
+[
+  (assignment (NAME))
+  (alias (NAME))
+  (value (NAME))
+  (parameter (NAME))
+] @variable
+
+; Recipe definitions
+[
+  (recipeheader (NAME))
+  (dependency (NAME))
+  (dependency (depcall (NAME)))
+] @function
+(parameter) @parameter
+(depcall (expression (value (NAME) @parameter)))
+
+; Fallback highlighting for recipe bodies
+(recipe
+  (body) @text.literal.block (#set! "priority" 90))
+(shebang_recipe
+  (shebang interpreter: (TEXT) @keyword) @comment
+  (shebang_body) @text.literal.block (#set! "priority" 90))
+
+; Ref: https://just.systems/man/en/chapter_26.html
+(setting (NAME) @error)
+(setting (NAME) @constant.builtin
+  (#any-of? @constant.builtin 
+    "allow-duplicate-recipes"
+    "dotenv-filename"
+    "dotenv-load"
+    "dotenv-path"
+    "export"
+    "fallback"
+    "ignore-comments"
+    "positional-arguments"
+    "shell"
+    "tempdir"
+    "windows-powershell"
+    "windows-shell"))
```

Comment:
Is this usage of `@error` acceptable?

Reply:
Well this seems to be very explicit, I'm okay with it.

Reply:
`@error` is specifically a parsing error, so it shouldn't in general be used at all. The goal of this(!) project is syntax highlighting, not linting -- this means we want to correctly highlight all _correct_ syntax, but highlighting all _incorrect_ syntax is an explicit non-goal-

Reply:
> `@error` is specifically a parsing error, so it shouldn't in general be used at all. The goal of this(!) project is syntax highlighting, not linting -- this means we want to correctly highlight all _correct_ syntax, but highlighting all _incorrect_ syntax is an explicit non-goal.

That's fair, but what's the best way to make users aware of typos like this (short of LSP)?

Should the treesitter parser fail to parse function- and setting names that don't exist, at which point they would be automatically highlighted as errors in Neovim?

Reply:
ask your parser maintainer to add that

Reply:
You're always welcome to put these captures in your own personal queries. They just don't fit in _our_ curated, consistent, queries.

Reply:
> The goal of this(!) project is syntax highlighting, not linting -- this means we want to correctly highlight all correct syntax, but highlighting all incorrect syntax is an explicit non-goal.

Sorry for harping on this but I still don't fully understand.

Following the goals of this project, what is the correct course of action when you're writing highlight queries for a parser that is "too nice", i.e. successfully parses files even if they are technically invalid?

I feel like highlighting such files naively is actively unhelpful, as it helps mask obvious errors, such as typos in keywords, indentation errors in Python code, etc.

Simultaneously, it's *much* easier and also more performant to write parsers that are a bit too nice, since properly handling errors in tree-sitter without temporarily borking the entire document can be very difficult.

Obvious example, Python files are parsed just fine with invalid indentation (a syntax error):

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/597206/0c81df1f-376c-42e3-b50b-c763b0ad1530)

It seems like pointing out certain syntax errors is just way easier with syntax highlighting, which this project has moved away from. (I still don't understand why, but I'm sure it's documented in the issues somewhere, I'll try to read up on it.)

Does nvim-treesitter need a new linter module? :smile:

Reply:
From reading this thread: https://github.com/nvim-treesitter/nvim-treesitter/pull/5572#issuecomment-1779113178

It seems like the consensus is that capturing `(ERROR)` nodes with `@error` is a bad thing, for several reasons that make a lot of sense. It sounds like applying `@error` to very specific captures that are *definitely* errors is still purely positive though, even if it's not the goal of the project :thinking:

What's the argument for strictly not allowing the `@error` capture anywhere?

Reply:
> Following the goals of this project, what is the correct course of action when you're writing highlight queries for a parser that is "too nice", i.e. successfully parses files even if they are technically invalid?

Do something outside the scope of this project.

> What's the argument for strictly not allowing the @error capture anywhere?

As I stated: Stick to one thing, and do that (as) well (as possible). For us, that thing is syntax highlighting, _not_ linting. 

> Does nvim-treesitter need a new linter module? 😄

Feel free to write a tree-sitter based linter. (An "nvim-treesitter" module is not welcome; we are actively working on getting rid of them.)

Reply:
The argument here is moreso: Let the parser handle the heavyweight of finding the error. We don't need/want to do it here. Users can add the reverse of your queries (`((NAME) @error (#not-eq? @error "..."))`) if they want error highlighting

Reply:
> > Following the goals of this project, what is the correct course of action when you're writing highlight queries for a parser that is "too nice", i.e. successfully parses files even if they are technically invalid?
> 
> Do something outside the scope of this project.

What I mean is, what is the correct course of action when writing the highlight queries?

- Should I highlight code that is *obviously* wrong as if it was correct?
  - I feel like this is actively harmful, people have an expectation that if it highlights as a keyword, it's a keyword
- Should I highlight code that is correct, while "leaving out" code that is wrong?
  - This feels like it's almost exactly the same as just capturing it with `@error`, but less helpful. At least if it was captured, users could choose to highlight it.

Are there any modules or plugins you know of that replaces the error highlighting that used to be in nvim-treesitter highlights? Personally I found it generally helpful to get live feedback while typing, without waiting for- and reading an error message from the LSP, assuming what I'm writing even has an LSP.

Reply:
> Should I highlight code that is correct, while "leaving out" code that is wrong?

Yes. (The big problem with highlighting syntax errors is that with treesitter, you get live feedback as you type. And (almost) any incomplete statement will be a syntax error, so you get constant blinking Christmas lights while typing. This is _horrible_ UX and not acceptable to us. And since consistency across queries is a primary project goal, this means _no_ query will highlight errors.)

> Are there any modules or plugins you know of that replaces the error highlighting that used to be in nvim-treesitter highlights? 

No. But you can do so very easily in your own config by adding queries that start with `; extends`.

Reply:
Fair enough, I can see how that would be really annoying with some color schemes.

> No. But you can do so very easily in your own config by adding queries that start with ; extends.

Right :thinking: So then it would be technically possible to write a plugin that extends all the highlights queries in nvim-treesitter to add the error captures back?

Reply:
Of course.